snapcraft/internal/steps.py

```diff
@@ -0,0 +1,102 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2018 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from typing import List
+
+
+class Step:
+    def __init__(self, name: str, clean_if_dirty: bool) -> None:
+        self.name = name
+        self.clean_if_dirty = clean_if_dirty
+        self.__order = None  # type: int
+
+    @property
+    def _order(self) -> int:
+        if self.__order is None:
+            self.__order = STEPS.index(self)
+        return self.__order
+
+    def previous_step(self) -> 'Step':
+        if self._order > 0:
+            return STEPS[self._order-1]
+        else:
+            return None
+
+    def next_step(self) -> 'Step':
+        try:
+            return STEPS[self._order+1]
+        except IndexError:
+            return None
+
+    def previous_steps(self) -> List['Step']:
+        return STEPS[:self._order]
+
+    def next_steps(self) -> List['Step']:
+        return STEPS[self._order+1:]
+
+    def __lt__(self, other) -> bool:
+        if type(other) is type(self):
+            return self._order < other._order
+
+        return NotImplemented
+
+    def __le__(self, other) -> bool:
+        if type(other) is type(self):
+            return self._order <= other._order
+
+        return NotImplemented
+
+    def __eq__(self, other) -> bool:
+        if type(other) is type(self):
+            return self.name == other.name
+
+        return NotImplemented
+
+    def __gt__(self, other) -> bool:
+        if type(other) is type(self):
+            return self._order > other._order
+
+        return NotImplemented
+
+    def __ge__(self, other) -> bool:
+        if type(other) is type(self):
+            return self._order >= other._order
+
+        return NotImplemented
+
+    def __hash__(self):
+        return hash(self.name)
+
+    def __repr__(self):
+        return 'snapcraft.internal.steps.Step({!r})'.format(self.name)
```

Comment:
It would of been nice to have some sort of `__modpath__` to not hardcode this, closest I know of is `__file__`, but it will look too hacky. If we ever move this module, this change will most likely go unnoticed, so why not remove the leading qualifiers?

Reply:
Yeah, you're right. I'm mostly going off of [python docs](https://docs.python.org/3/reference/datamodel.html#object.__repr__) here that say:

> [...] If at all possible, this should look like a valid Python expression that could be used to recreate an object with the same value

So I added the qualifiers. I agree that it's just asking to break though, and I question its usefulness anyway. This is really just to make test failures nice, so we can make it whatever we want. I'll remove the leading qualifiers.

Reply:
Done.